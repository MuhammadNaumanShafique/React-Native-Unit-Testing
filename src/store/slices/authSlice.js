import {createSlice, createAsyncThunk} from '@reduxjs/toolkit';
import Api from '../../services/Api';

export const loginUser = createAsyncThunk(
  'auth/loginUser',
  async (params, thunkAPI) => {
    const api = Api.create();
    // const loginResponse = await api
    //   .userLoginApi(params)
    //   .then(data => data)
    //   .catch(error => thunkAPI.rejectWithValue(error?.data || error));
    // return loginResponse;
    return {};
  },
);
// export const signupUser = createAsyncThunk(
//   'auth/signupUser',
//   async (params, thunkAPI) => {
//     thunkAPI.dispatch(authSlice.actions.addUser(params));
//   },
// );

// export const signupUser = createAsyncThunk(
//   'auth/signupUser',
//   async (params, thunkAPI) => {
//     console.log('signup api called');
//     const api = Api.create();
//     const signupResponse = await api
//       .userSignupApi(params)
//       .then(data => {
//         console.log('response found: ', data);
//         return data;
//       })
//       .catch(error => {
//         console.log('Error found: ', error);
//         return thunkAPI.rejectWithValue(error?.data || error);
//       });

//     console.log('here reached');
//     return signupResponse;
//   },
// );

const authSlice = createSlice({
  name: 'auth',
  initialState: {
    users: [],
    userData: '',
    loading: false,
  },
  reducers: {
    addUser: (state, payload) => {
      state.users.push(payload.payload);
    },
    logout: state => {
      state.loading = false;
    },
    resetLoader: state => {
      state.loading = false;
    },
  },
  // extraReducers: builder => {
  //   builder.addCase(signupUser.fulfilled, (state, action) => {
  //     console.log('data received: ', action.payload);
  //     state.users.push(action.payload);
  //   });
  // },
});

export const {logout, addUser} = authSlice.actions;

export default authSlice.reducer;
