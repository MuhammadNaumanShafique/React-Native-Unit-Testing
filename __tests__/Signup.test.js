import {render, fireEvent} from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import {Provider} from 'react-redux';
import configureStore from 'redux-mock-store';
import SignUpScreen from '../src/screens/signup/Signup';
import {addUser} from '../src/store/slices/authSlice';

jest.mock('@react-native-async-storage/async-storage', () => ({
  getItem: jest.fn(),
  setItem: jest.fn(),
  removeItem: jest.fn(),
}));

const mockNavigate = jest.fn();
const mockGoBack = jest.fn();

jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useNavigation: () => ({
    navigate: mockNavigate,
    goBack: mockGoBack,
  }),
}));

describe('SignupScreen component', () => {
  const mockStore = configureStore([]);

  const store = mockStore({
    auth: {users: [{email: 'nauman@gmail.com', password: 'password123'}]},
  });

  it('SignupScreen renders correctly', async () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <SignUpScreen />,
        </Provider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders email and password inputs', () => {
    const {getByPlaceholderText} = render(
      <Provider store={store}>
        <SignUpScreen />,
      </Provider>,
    );
    const emailInput = getByPlaceholderText('Enter your email');
    const passwordInput = getByPlaceholderText('Password');

    expect(emailInput).toBeTruthy();
    expect(passwordInput).toBeTruthy();
  });

  it('updates email and password state on input change', () => {
    const {getByPlaceholderText} = render(
      <Provider store={store}>
        <SignUpScreen />,
      </Provider>,
    );
    const emailInput = getByPlaceholderText('Enter your email');
    const passwordInput = getByPlaceholderText('Password');

    fireEvent.changeText(emailInput, 'test@example.com');
    fireEvent.changeText(passwordInput, 'password123');

    expect(emailInput.props.value).toBe('test@example.com');
    expect(passwordInput.props.value).toBe('password123');
  });

  it('executes loginUser function on button press with existing credentials', () => {
    const {getByText, getByPlaceholderText} = render(
      <Provider store={store}>
        <SignUpScreen />,
      </Provider>,
    );
    const signupButton = getByText('Signup');
    const consoleSpy = jest.spyOn(console, 'log');

    const emailInput = getByPlaceholderText('Enter your email');
    const passwordInput = getByPlaceholderText('Password');

    fireEvent.changeText(emailInput, 'nauman@gmail.com');
    fireEvent.changeText(passwordInput, 'password123');

    expect(emailInput.props.value).toBe('nauman@gmail.com');
    expect(passwordInput.props.value).toBe('password123');

    fireEvent.press(signupButton);

    expect(consoleSpy).toHaveBeenCalledWith('User already exists try sign in');

    consoleSpy.mockRestore();
  });

  it('executes signup function on button press with new credentials', () => {
    const {getByTestId} = render(
      <Provider store={store}>
        <SignUpScreen />
      </Provider>,
    );

    const emailInput = getByTestId('emailInputField');
    const passwordInput = getByTestId('passwordInputField');
    const signupButton = getByTestId('signupButton');

    fireEvent.changeText(emailInput, 'test@example.com');
    fireEvent.changeText(passwordInput, 'password');
    fireEvent.press(signupButton);

    const actions = store.getActions();
    const expectedAction = addUser({
      email: 'test@example.com',
      password: 'password',
    });

    expect(actions).toContainEqual(expectedAction);
    expect(actions.length).toBe(1); // Ensure only one action is dispatched
    expect(mockNavigate).toHaveBeenCalledWith('loginScreen');
  });
});
