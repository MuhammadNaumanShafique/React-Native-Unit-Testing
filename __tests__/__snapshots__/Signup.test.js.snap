// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupScreen component SignupScreen renders correctly 1`] = `
[
  <RCTSafeAreaView>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "column",
          "height": "100%",
          "justifyContent": "center",
          "marginHorizontal": 30,
        }
      }
    >
      <Text
        style={
          {
            "fontSize": 30,
            "marginBottom": "20%",
            "textAlign": "center",
          }
        }
      >
        Sign up
      </Text>
      <TextInput
        onChangeText={[Function]}
        placeholder="Enter your email"
        style={
          {
            "borderColor": "#ccc",
            "borderWidth": 2,
            "marginBottom": 20,
            "padding": 20,
            "width": "100%",
          }
        }
        testID="emailInputField"
        value=""
      />
      <TextInput
        onChangeText={[Function]}
        placeholder="Password"
        style={
          {
            "borderColor": "#ccc",
            "borderWidth": 2,
            "marginBottom": 20,
            "padding": 20,
            "width": "100%",
          }
        }
        testID="passwordInputField"
        value=""
      />
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="signupButton"
      >
        <View
          style={
            [
              {},
            ]
          }
        >
          <Text
            style={
              [
                {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
                {
                  "color": "blue",
                },
              ]
            }
          >
            Signup
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "flexDirection": "row",
            "marginTop": 20,
          }
        }
      >
        <Text>
          Already have an account?
        </Text>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              {
                "color": "blue",
              }
            }
          >
             Sign in
          </Text>
        </View>
      </View>
    </View>
  </RCTSafeAreaView>,
  ",",
]
`;
